### DISCUSSION TITLE
Adding Custom Controller  to DGS graphql framework

### DISCUSSION THREAD

**Discussion by madhukanth2005** [Netflix/dgs-framework#286] [Q&A]
Hi,
I am new to DGS netflix and we are developing an application Using Spring boot with DGS Netflix graphql. We have created all the schemas and datafethers which is working absolutely fine with a default endpoint "/graphql". we would like to expose this app with custom endpoing and write some common code  ,so we are trying to add a controller with a custom endpoint as below (we need to have lot of other logic in controller). But When i run the application and send a query, my data fetcher being called twice . first time called from controller (I think from dgsQueryExecutor.execute() method )and second time i believe from dgs framework it self. Anybody has any thoughts on this why its being called twice and how to avoid it? Also , i would like to set a customContext object while executing (in my case its modelGraphQLContext). Please see the below Datafetcher and Controller for your reference.Your help is highly appreciated. Thank you in advance.

Controller:

@RestController
@RequestMapping("/sample-information/model")
@Slf4j
public class CustomController {
    
    @Autowired
    DgsQueryExecutor dgsQueryExecutor;

    
    @PostMapping(consumes = {MediaType.APPLICATION_JSON_VALUE, "application/graphql"})
    public Mono<ResponseEntity<Object>> getDetails(@RequestBody String query,
            @RequestHeader HttpHeaders headers
            ) {
        GraphQLQueryInput inputs = null;
        try {
            inputs = ObjectMapperHelper.objectMapper.readValue(query, GraphQLQueryInput.class);
        } catch (Exception e) {
            log.error("TraceId: {} - Application Error: Error message: Error converting query to GraphQLQueryInput: {} "+ query);
            
        }
        
        if(inputs.getVariables() == null) {
            inputs.setVariables(new HashMap<>());
        }
        
        if(inputs.getOperationName() == null) {
            inputs.setOperationName("");
        }
        
        final String que = inputs.getQuery();
        final Map<String, Object> var = inputs.getVariables();
        final String opn = inputs.getOperationName();
        
        ExecutionInput.Builder executionInput = ExecutionInput.newExecutionInput()
                .query(inputs.getQuery())
                .operationName(inputs.getOperationName())
                .variables(inputs.getVariables());
        
       return Mono.fromCallable(()-> {
            return dgsQueryExecutor.execute(que, var, opn);
        }).subscribeOn(Schedulers.elastic()).map(result -> {
            return new ResponseEntity<>(result, HttpStatus.OK); 
        }); 
    }

}

Datafetcher:

@DgsComponent
@Slf4j
public class SampleDataFetcher {
    @Autowired
    SampleBuilder sampleBuilder;
    
    @DgsData(parentType = DgsConstants.QUERY_TYPE, field = DgsConstants.QUERY.SampleField)
    public CompletableFuture<StoreDirectoryByStateResponse> getStoreDirectoryByState(@InputArgument String state, DataFetchingEnvironment dfe) throws BadRequestException {

        Mono<StoreDirectoryByStateResponse> storeDirectoryResponse = null;
        try {
            sampleResponse = sampleBuilder.buildResponse(modelGraphQLContext);
        } catch (BaseException e) {

        }
        return sampleResponse.map(response -> {
            return response;
        }).toFuture();
        
    }
}

**Comment by paulbakker** [comment]
What are you doing with this code? It doesn't seem to be used at all? 

```
 ExecutionInput.Builder executionInput = ExecutionInput.newExecutionInput()
            .query(inputs.getQuery())
            .operationName(inputs.getOperationName())
            .variables(inputs.getVariables());
```

Other than that, I see no reason why the datafetcher would be called twice. You can try set a breakpoint in this controller, and double check the controller itself isn't called twice for some reason. You could also set a breakpoint in the framework's `DgsRestController` to double check it doesn't also get called (it shouldn't).

Conceptually, what you're trying to do should work.

**Reply by madhukanth2005** [reply]
Thanks Paul for the quick response.the above mentioned code in your comments not being used. I was trying to set my custom context object and execute but i did not find a way to do so. as mentioned below for reference. I have tested multiple times by keeping the break point in controller and datafetcher. i see controller being called only once but datafetcher being called twice. looks like when the line dgsQueryExecutor.execute(que, var, opn) executes in the controlled datafetcher is called and secod time directly calls the datafetcher with out calling controller. I also notice the logs in the console are printed twice.
 
ExecutionInput.Builder executionInput = ExecutionInput.newExecutionInput()
                .query(inputs.getQuery())
                .operationName(inputs.getOperationName())
                .variables(inputs.getVariables());
        executionInput.context(modelGraphQLContext);
        dgsQueryExecutor.execute(executionInput); //This gives a compilation error as execute() method won't take //executionInput as parameter

**Reply by madhukanth2005** [reply]
Also, Could you please help me how to pass the custom context object in the execute method? i know how to pass in java graphql as execute method takes executionInput as param as shown below. But i am not sure how to achieve the same in dgs netflix. Your help is highly appreciated as this has become a blocker for us. Thanks in advance.


executionInput.context(modelGraphQLContext);
graphQL.execute(executionInput);
